src/language_features/kak_lsp_custom.rs

```diff
@@ -0,0 +1,318 @@
+    editor_params: EditorParams,
+    result: Option<DocumentSymbolResponse>,
+    ctx: &mut Context,
+) {
+    let params = NextOrPrevSymbolParams::deserialize(editor_params).unwrap();
+    let hover = params.hover;
+    let maybe_details = match result {
+        Some(DocumentSymbolResponse::Flat(mut result)) => {
+            if result.is_empty() {
+                return;
+            }
+            // First let's sort the results so we can find next and previous properly
+            // This step does _not_ happen in `document_symbol::editor_document_symbol()`.
+            //
+            // Some language servers return symbol locations in unsorted order
+            // or non-ascending order.
+            result.sort_by(|a, b| a.location.range.start.cmp(&b.location.range.start));
+            get_next_or_prev_symbol_information_details(result, params, ctx)
+        }
+        Some(DocumentSymbolResponse::Nested(mut result)) => {
+            if result.is_empty() {
+                return;
+            }
+            // First let's sort the results so we can find next and previous properly
+            // This step does _not_ happen in `document_symbol::editor_document_symbol()`.
+            //
+            // Some language servers return symbol locations in unsorted order
+            // or non-ascending order.
+            result.sort_by(|a, b| a.range.start.cmp(&b.range.start));
+            get_next_or_prev_document_symbol_details(result, params, &meta, ctx)
+        }
+        None => {
+            return;
+        }
+    };
+
+    next_or_prev_response(meta, ctx, maybe_details, hover);
+}
+
+/// Send the response back to Kakoune. This could be either:
+/// a) Instructions to move the cursor to the next/previous symbol.
+/// b) Instructions to show hover information (without actually moving the
+///    visible user cursor) _of_ the next/previous symbol
+fn next_or_prev_response(
+    meta: EditorMeta,
+    ctx: &mut Context,
+    maybe_details: Option<(String, KakounePosition, String, SymbolKind)>,
+    hover: bool,
+) {
+    if let Some((filename, symbol_position, name, kind)) = maybe_details {
+        if hover {
+            // Create hover `req_params` just like we do in `hover::text_document_hover()`
+            let req_params = HoverParams {
+                text_document_position_params: TextDocumentPositionParams {
+                    text_document: TextDocumentIdentifier {
+                        uri: Url::from_file_path(&meta.buffile).unwrap(),
+                    },
+                    position: get_lsp_position(&meta.buffile, &symbol_position, ctx).unwrap(),
+                },
+                work_done_progress_params: Default::default(),
+            };
+
+            // This context is shown at the top of the modal
+            let context = format!(
+                                "line {}:{}:{{+b@KindAndName}}{:?} {}{{KindAndName}} (Press `g` to goto this position. Press any other key to continue)",
```

Comment:
The docstring of Kakoune's `add-highlighter` uses single quotes, which fits better than backticks IMO. Or maybe no quotes at all?

```suggestion
                                "line {}:{}:{{+b@KindAndName}}{:?} {}{{KindAndName}} (Press 'g' to goto this position. Press any other key to continue)",
```